File: problems/CSES/Sorting_and_Searching/movie_festival_ii/movie_festival_ii.py
import sys
from bisect import bisect_right
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

class FenwickTree:
    def __init__(self, n):
        self.n = n
        self.bit = [0] * (n + 1)

    def add(self, i, value):
        while i <= self.n:
            self.bit[i] += value
            i += i & -i

    def subtract(self, i, value):
        while i <= self.n:
            self.bit[i] -= value
            i += i & -i

    def prefix_sum(self, i):
        prefix_sum = 0

        while i > 0:
            prefix_sum += self.bit[i]
            i -= i & -i

        return prefix_sum

    def range_sum(self, start, end):
        return self.prefix_sum(end) - self.prefix_sum(start - 1)

    def kth(self, k):
        index = 0
        step = 1 << (self.n.bit_length() - 1)

        while step:
            next_index = index + step

            if next_index <= self.n and self.bit[next_index] < k:
                index = next_index
                k -= self.bit[next_index]

            step >>= 1

        return index + 1

def main():
    n, k = map(int, input().split())
    movies = []
    times = [0]

    for _ in range(n):
        start, finish = map(int, input().split())
        movies.append((start, finish))
        times.append(finish)

    movies.sort(key=lambda movie: movie[1])

    times = sorted(set(times))

    fenwick = FenwickTree(len(times))

    zero_index = 1
    fenwick.add(zero_index, k)

    res = 0

    for start, finish in movies:
        position = bisect_right(times, start)

        available_count = fenwick.prefix_sum(position)

        if available_count == 0:
            continue

        member_index = fenwick.kth(available_count)

        fenwick.add(member_index, -1)

        finish_index = bisect_right(times, finish)
        fenwick.add(finish_index, 1)

        res += 1

    print(res)

main()

File: problems/CSES/Sorting_and_Searching/movie_festival_ii/movie_festival_ii.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

class FenwickTree {
private:
    int n;
    vector<int> bit;

public:
    explicit FenwickTree(int size)
        : n(size), bit(size + 1, 0) {}

    void add(int index, int value) {
        while (index <= n) {
            bit[index] += value;
            index += index & -index;
        }
    }

    void subtract(int index, int value) {
        add(index, -value);
    }

    int prefixSum(int index) const {
        int sum = 0;

        while (index > 0) {
            sum += bit[index];
            index -= index & -index;
        }

        return sum;
    }

    int rangeSum(int left, int right) const {
        return prefixSum(right) - prefixSum(left - 1);
    }

    int kth(int k) const {
        int index = 0;

        int step = 1;
        while ((step << 1) <= n) {
            step <<= 1;
        }

        while (step > 0) {
            int nextIndex = index + step;

            if (nextIndex <= n && bit[nextIndex] < k) {
                index = nextIndex;
                k -= bit[nextIndex];
            }

            step >>= 1;
        }

        return index + 1;
    }
};

int main_v1() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, k;
    cin >> n >> k;

    vector<pair<int, int>> movies;
    movies.reserve(n);

    vector<int> times;
    times.reserve(n + 1);
    times.push_back(0);

    for (int i = 0; i < n; i++) {
        int start, finish;
        cin >> start >> finish;

        movies.emplace_back(start, finish);
        times.push_back(finish);
    }

    sort(movies.begin(), movies.end(),
         [](const pair<int, int>& a, const pair<int, int>& b) {
             return a.second < b.second;
         });

    sort(times.begin(), times.end());
    times.erase(unique(times.begin(), times.end()), times.end());

    FenwickTree fenwick(static_cast<int>(times.size()));

    int zeroIndex =
        static_cast<int>(upper_bound(times.begin(), times.end(), 0)
                         - times.begin());

    fenwick.add(zeroIndex, k);

    int result = 0;

    for (auto [start, finish] : movies) {
        int position =
            static_cast<int>(upper_bound(times.begin(), times.end(), start)
                             - times.begin());

        int availableCount = fenwick.prefixSum(position);

        if (availableCount == 0) {
            continue;
        }

        int memberIndex = fenwick.kth(availableCount);

        fenwick.add(memberIndex, -1);

        int finishIndex =
            static_cast<int>(lower_bound(times.begin(), times.end(), finish)
                             - times.begin()) + 1;

        fenwick.add(finishIndex, 1);

        result++;
    }

    cout << result << '\n';

    return 0;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, k;
    cin >> n >> k;

    vector<pair<int, int>> movies(n);
    for (int i = 0; i < n; i++)
        cin >> movies[i].second >> movies[i].first;

    sort(movies.begin(), movies.end());

    multiset<int> ends;
    for (int i = 0; i < k; i++)
        ends.insert(0);

    int res = 0;
    for (auto [finish, start] : movies) {
        auto it = ends.upper_bound(start);

        if (it == ends.begin())
            continue;

        it--;
        ends.erase(it);
        ends.insert(finish);
        res++;
    }

    cout << res << '\n';

    return 0;
}
